services/galley/src/Galley/API/Update.hs

```diff
@@ -333,73 +326,71 @@ joinConversation zusr zcon cnv access = do
         ensureAccessRole (Data.convAccessRole conv) newUsers (Just tms)
-        void $ permissionCheck zusr AddRemoveConvMember tms
         tcv <- Data.teamConversation tid cid
         when (maybe True (view managedConversation) tcv) $
             throwM noAddToManaged
-        -- Team members are alwasy considered connected, so we only check 'ensureConnected'
+        -- Team members are always considered connected, so we only check 'ensureConnected'
         -- for non-team-members.
         let guests = notTeamMember newUsers tms
         ensureConnected zusr guests
 
-updateMember :: UserId ::: ConnId ::: ConvId ::: JsonRequest MemberUpdate -> Galley Response
-updateMember (zusr ::: zcon ::: cid ::: req) = do
-    alive <- Data.isConvAlive cid
-    unless alive $ do
-        Data.deleteConversation cid
-        throwM convNotFound
+updateSelfMember :: UserId ::: ConnId ::: ConvId ::: JsonRequest MemberUpdate -> Galley Response
+updateSelfMember (zusr ::: zcon ::: cid ::: req) = do
+    conv <- getConversationAndCheckMembership zusr cid
     body <- fromJsonBody req
-    m    <- Data.member cid zusr >>= ifNothing convNotFound
-    up   <- Data.updateMember cid zusr body
-    now  <- liftIO getCurrentTime
-    let e = Event MemberStateUpdate cid zusr now (Just $ EdMemberUpdate up)
-    let ms = applyChanges m up
-    for_ (newPush (evtFrom e) (ConvEvent e) [recipient ms]) $ \p ->
-        push1 $ p
-              & pushConn  ?~ zcon
-              & pushRoute .~ RouteDirect
+    m    <- getSelfMember zusr (Data.convMembers conv)
+
+    void $ processUpdateMemberEvent zusr zcon cid [m] m body
+    return empty
+
+updateOtherMember :: UserId ::: ConnId ::: ConvId ::: UserId ::: JsonRequest OtherMemberUpdate -> Galley Response
+updateOtherMember (zusr ::: zcon ::: cid ::: victim ::: req) = do
+    conv <- getConversationAndCheckMembership zusr cid
+    body <- fromJsonBody req
+    let (bots, users) = botsAndUsers (Data.convMembers conv)
+    ensureActionAllowed ModifyOtherConversationMember =<< getSelfMember zusr users
+    memTarget <- getOtherMember victim users
+
+    e <- processUpdateMemberEvent zusr zcon cid users memTarget (toMemberUpdate body)
+    void . forkIO $ void $ External.deliver (bots `zip` repeat e)
     return empty
   where
-    applyChanges :: Member -> MemberUpdateData -> Member
-    applyChanges m u = m
-        { memOtrMuted       = fromMaybe (memOtrMuted m) (misOtrMuted u)
-        , memOtrMutedRef    = misOtrMutedRef u <|> memOtrMutedRef m
-        , memOtrArchived    = fromMaybe (memOtrArchived m) (misOtrArchived u)
-        , memOtrArchivedRef = misOtrArchivedRef u <|> memOtrArchivedRef m
-        , memHidden         = fromMaybe (memHidden m) (misHidden u)
-        , memHiddenRef      = misHiddenRef u <|> memHiddenRef m
-        }
+    toMemberUpdate :: OtherMemberUpdate -> MemberUpdate
+    toMemberUpdate omu = MemberUpdate Nothing Nothing Nothing Nothing Nothing Nothing Nothing (omuConvRoleName omu)
```

Comment:
This is not pretty, can you use named fields here?

Reply:
Yes, I improved that 👍